query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+        .await?;
+
+        *inner = Inner::Connected(engine);
+        Ok(())
+    }
+
+    pub async fn query(
+        &self,
+        body_str: *const c_char,
+        trace_str: *const c_char,
+        tx_id_str: *const c_char,
+    ) -> Result<String, Error> {
+        let dispatcher = self.logger.dispatcher();
+
+        async move {
+            let inner = self.inner.read().await;
+            let engine = inner.as_engine()?;
+
+            let body = get_cstr_safe(body_str).expect("Prisma engine execute body is missing");
+            let tx_id = get_cstr_safe(tx_id_str);
+            let trace = get_cstr_safe(trace_str).expect("Trace is needed");
+
+            let query = RequestBody::try_from_str(&body, engine.engine_protocol()).map_err(Error::from)?;
+
+            let span = if tx_id.is_none() {
+                tracing::info_span!("prisma:engine", user_facing = true)
+            } else {
+                Span::none()
+            };
+
+            let trace_id = telemetry::helpers::set_parent_context_from_json_str(&span, &trace);
+
+            async move {
+                let handler = RequestHandler::new(engine.executor(), engine.query_schema(), engine.engine_protocol());
+                let response = handler.handle(query, tx_id.map(TxId::from), trace_id).await;
+
+                let serde_span = tracing::info_span!("prisma:engine:response_json_serialization", user_facing = true);
+                Ok(serde_span.in_scope(|| serde_json::to_string(&response))?)
+            }
+            .instrument(span)
+            .await
+        }
+        .with_subscriber(dispatcher)
+        .await
+    }
+
+    /// Disconnect and drop the core. Can be reconnected later with `#connect`.
+    pub async fn disconnect(&self, trace_str: *const c_char) -> Result<(), Error> {
+        let trace = get_cstr_safe(trace_str).expect("Trace is needed");
+        // let dispatcher = self.logger.dispatcher();
+
+        // async_panic_to_js_error(async {
+        let span = tracing::info_span!("prisma:engine:disconnect");
+        let _ = telemetry::helpers::set_parent_context_from_json_str(&span, &trace);
+
+        // TODO: when using Node Drivers, we need to call Driver::close() here.
+
+        // async {
+        let mut inner = self.inner.write().await;
+        let engine = inner.as_engine()?;
+
+        let builder = EngineBuilder {
+            schema: engine.schema.clone(),
+            engine_protocol: engine.engine_protocol(),
+            native: EngineBuilderNative {
+                config_dir: engine.native.config_dir.clone(),
+                env: engine.native.env.clone(),
+            },
+        };
+
+        *inner = Inner::Builder(builder);
+
+        Ok(())
+        // }
+        // .instrument(span)
+        // .await
+        // })
+        // .with_subscriber(dispatcher)
+        // .await
+    }
+
+    async unsafe fn apply_migrations(&self, migration_folder_path: *const c_char) -> Result<(), Error> {
```

Comment:
Yup definitely, at the same time, the use case will be different. Handling errors will be done differently on thousands of devices at the same time. While on the server the person doing the migration has hands-on keyboard, on mobile this will need to have a lot more failsafe to allow apps to start in case of a failed migration.

I haven't completely thought through was this means since this is a completely new use case for Prisma.

For now I will add the TODO but it might actually make sense to keep it separated.